### PR TITLE
Add support of Sparse Tensor Slice

### LINKED ISSUE #1588: Slicing sparse tensor
There does not appear to be a way to extract a (sparse or dense) slice from a sparse tensor in the same manner as there is from a dense tensor. Will this be a feature in the future, or is it already present and I'm just missing something? Perhaps it should be added to the planned improvements for indexing.

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #1588 to support slice of sparse tensors.

This fix fixes #1588.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/core/util/sparse/sparse_tensor.h [hunk 1]
```diff
@@ -148,6 +148,13 @@ class SparseTensor {
                                          const int split_dim,
                                          const int num_split);
 
+  // Slice() will slice the input SparseTensor into a SparseTensor based on
```
Comment (ebrevdo) [inline tensorflow/core/util/sparse/sparse_tensor.h]: document expected sizes of start & size and what they refer to inside the ST.

File: tensorflow/core/util/sparse/sparse_tensor.h [hunk 2]
```diff
@@ -148,6 +148,13 @@ class SparseTensor {
...
+  // specified start and size.
+  template <typename T>
+  static SparseTensor Slice(const SparseTensor& tensor,
+                            const gtl::ArraySlice<int64> &start,
```
Comment (ebrevdo) [inline tensorflow/core/util/sparse/sparse_tensor.h]: TF's code guidelines require the amperstand to be part of the type:  const gtl::ArraySlice<int64>& start

File: tensorflow/core/util/sparse/sparse_tensor.h [hunk 3]
```diff
@@ -530,6 +537,65 @@ std::vector<SparseTensor> SparseTensor::Split(const SparseTensor& input_tensor,
...
 
+template <typename T>
+SparseTensor SparseTensor::Slice(const SparseTensor& input_tensor,
+                                 const gtl::ArraySlice<int64> &start,
+                                 const gtl::ArraySlice<int64> &size) {
+  TensorShape output_shape = input_tensor.shape();
+
+  const int dims = input_tensor.shape().dims();
+  for (int dim = 0; dim < dims; dim++) {
+    int64 dim_size = start[dim] + size[dim] < output_shape.dim_size(dim) ? size[dim] : output_shape.dim_size(dim) - start[dim];
```
Comment (ebrevdo) [inline tensorflow/core/util/sparse/sparse_tensor.h]: please run clang-format on your code.  this is above the 80 char limit.

File: tensorflow/core/util/sparse/sparse_tensor.h [hunk 4]
```diff
@@ -530,6 +537,65 @@ std::vector<SparseTensor> SparseTensor::Split(const SparseTensor& input_tensor,
...
+  }
+
+  auto input_indices_t = input_tensor.indices().matrix<int64>();
+  auto input_values_t = input_tensor.values().vec<T>();
+
+  int count = 0;
+  for (int i = 0; i < input_tensor.indices().dim_size(0); i++) {
+    bool hit = true;
+    for (int dim = 0; dim < dims; dim++) {
+      if (!(start[dim] <= input_indices_t(i, dim) && input_indices_t(i, dim) < start[dim] + size[dim])) {
```
Comment (ebrevdo) [inline tensorflow/core/util/sparse/sparse_tensor.h]: are you making assumptions about the sorting of the SparseTensor?  Do you need to run ValidateIndices first?

File: tensorflow/core/util/sparse/sparse_tensor.h [hunk 5]
```diff
@@ -530,6 +537,65 @@ std::vector<SparseTensor> SparseTensor::Split(const SparseTensor& input_tensor,
...
+
+  auto output_values_t = output_values.vec<T>();
+  auto output_indices_t = output_indices.matrix<int64>();
+
+  int index = 0;
+  for (int i = 0; i < input_tensor.indices().dim_size(0) && index < count; i++) {
+    bool hit = true;
+    
+    for (int dim = 0; dim < dims; dim++) {
+      if (!(start[dim] <= input_indices_t(i, dim) && input_indices_t(i, dim) < start[dim] + size[dim])) {
```
Comment (ebrevdo) [inline tensorflow/core/util/sparse/sparse_tensor.h]: add some comments about what you're doing.

File: tensorflow/core/util/sparse/sparse_tensor.h [hunk 6]
```diff
@@ -530,6 +537,65 @@ std::vector<SparseTensor> SparseTensor::Split(const SparseTensor& input_tensor,
   return output_tensors;
 }
 
+template <typename T>
+SparseTensor SparseTensor::Slice(const SparseTensor& input_tensor,
```
Comment (ebrevdo) [inline tensorflow/core/util/sparse/sparse_tensor.h]: can you add a C++ test?  i believe there's a util/sparse/sparse_tensor_test.cc where you can add one.

**Review by ebrevdo — DISMISSED** [review]

**Comment by josh11b** [comment]
This sounds good for API review.

**Comment by rmlarsen** [comment]
@yongtang Whendo you think you can address @ebrevdo's comments?

File: tensorflow/core/util/sparse/sparse_tensor.h (hunk 1 above)
Comment (yongtang) [inline tensorflow/core/util/sparse/sparse_tensor.h]: Thanks @ebrevdo. The PR has been updated with comments added.

File: tensorflow/core/util/sparse/sparse_tensor.h (hunk 2 above)
Comment (yongtang) [inline tensorflow/core/util/sparse/sparse_tensor.h]: Done.

File: tensorflow/core/util/sparse/sparse_tensor.h (hunk 6 above)
Comment (yongtang) [inline tensorflow/core/util/sparse/sparse_tensor.h]: Thanks. Done.

File: tensorflow/core/util/sparse/sparse_tensor.h (hunk 3 above)
Comment (yongtang) [inline tensorflow/core/util/sparse/sparse_tensor.h]: Thanks. Done.

File: tensorflow/core/util/sparse/sparse_tensor.h (hunk 4 above)
Comment (yongtang) [inline tensorflow/core/util/sparse/sparse_tensor.h]: @ebrevdo There is no assumption here. The logic is to find the number of output indices. I updated the comments. Please take a look.

File: tensorflow/core/util/sparse/sparse_tensor.h (hunk 5 above)
Comment (yongtang) [inline tensorflow/core/util/sparse/sparse_tensor.h]: Thanks @ebrevdo. The PR has been updated with added comments.

**Comment by yongtang** [comment]
@ebrevdo @josh11b @vrv @rmlarsen Thanks for the review and sorry for the delay. The PR has been updated to address the comments. Please take a look and let me know if there are any issues.

**Comment by rmlarsen** [comment]
Looks good to me, @ebrevdo can you take another look and approve?

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by yongtang** [comment]
@martinwicke The Jenkins build error seems to be for real. Let me take a look and see if I could fix it.

**Comment by yongtang** [comment]
@martinwicke The PR has been updated with build error fixed. Please take a look.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by yongtang** [comment]
@martinwicke There was another Jenkins error so the PR has been updated again. Now all the tests should pass. Sorry for the back and forth updates in the last several PR updates.

**Comment by martinwicke** [comment]
Not your fault. Thanks for updating!

Jenkins, test this please.

**Comment by yongtang** [comment]
@martinwicke The PR has been updated with API goldens updater so that `//tensorflow/tools/api/tests:api_compatibility_test` passes. Please take a look.

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by drpngx** [comment]
It looks like @martinwicke approved, so good to go?

**Comment by martinwicke** [comment]
Good to go.

**Pull request merged**